### PR TITLE
Cpr 146

### PR DESCRIPTION
**Pull request by vb-moj**
Added PNC mismatch track event.

Refactored the code to make it more readable and separate the matcher from the service.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt [hunk 1]
```diff
@@ -28,40 +28,65 @@ class OffenderService(
...
+      val offenderMatcher = getOffenderMatcher(person)
+      when {
+        offenderMatcher.isPncDoesNotMatch() -> pncDoesNotMatch(offenderMatcher, personEntity, person)
+        offenderMatcher.isExactMatch() -> exactMatchFound(offenderMatcher, personEntity, person)
+        offenderMatcher.isPartialMatch() -> partialMatchFound(personEntity, person)
+        offenderMatcher.isMultipleMatch() -> multipleMatchesFound(offenderMatcher, personEntity, person)
+        else -> noRecordsFound(personEntity, person)
+      }
+    }
+  }
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt:42]: Very nice, readable and has an else to catch all conditions.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt [hunk 2]
```diff
@@ -28,40 +28,65 @@ class OffenderService(
...
+    telemetryService.trackEvent(
+      TelemetryEventType.DELIUS_PNC_MISMATCH,
+      mapOf(
+        "UUID" to personEntity.personId.toString(),
+        "PNC searched for" to person.otherIdentifiers?.pncIdentifier?.pncId,
+        "PNC returned from search" to offenderDetails.joinToString(" ") { it.otherIds.pncNumber.toString() },
+        "PRISON NUMBER" to offenderDetails.singleOrNull()?.otherIds?.nomsNumber,
+      ).filterValues { !it.isNullOrBlank() },
+    )
+  }
```
Comment (vb-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt:89]: I think I can merge this with the other one. let me try something.

**Review by markreesmoj — APPROVED** [review]
👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderServiceTest.kt [hunk 3]
```diff
@@ -79,30 +79,13 @@ class OffenderServiceTest {
     )
   }
 
-  @Test
```
[the remark this answers is not in the document]
Comment (vb-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderServiceTest.kt:82]: We are not dealing with possible matches yet. And treating that as a No Match.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt (hunk 2 above)
Comment (vb-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt:89]: Merging makes the code unnecessarily messy by passing extra parameters without the need for it.

**Pull request merged**